File: Rapporteur/Pont.swift
import Foundation
import WebKit

/// Le seul pont entre la page et l'application — l'équivalent du preload
/// Electron et du `PontRapporteur` Android.
///
/// Côté page, `pont.js` (injecté au chargement) pose `window.rapporteurIOS`,
/// et `window.rapporteurAndroid` en alias : le site reconnaît ainsi une
/// « application mobile » sans rien changer à ses branches Android, et lit
/// `rapporteurIOS` là où l'iPhone diffère (capture native, textes).
///
/// Les appels sont asynchrones : la page poste `{id, action, params}`, et
/// l'application répond par `rapporteurIOS._repondre(id, réponse)` — les
/// données binaires (fichier, extrait) passent avant, par morceaux base64
/// (`_morceau`). Rien d'autre ne traverse : pas d'accès au système depuis la page.
final class Pont: NSObject, WKScriptMessageHandler {

    static let nom = "rapporteur"

    private weak var vue: WKWebView?
    private let enregistreur = Enregistreur()
    private var minuterieNiveau: Timer?
    /// Les gros transferts (le fichier de la séance) passent un par un.
    private let fileEnvois = DispatchQueue(label: "com.lerapporteur.mobile.pont")

    var version: String {
        (Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String) ?? "0"
    }

    func attacher(a vue: WKWebView) {
        self.vue = vue
        vue.configuration.userContentController.add(self, name: Pont.nom)
        if let chemin = Bundle.main.url(forResource: "pont", withExtension: "js"),
           let source = try? String(contentsOf: chemin, encoding: .utf8) {
            let script = WKUserScript(source: source.replacingOccurrences(of: "__VERSION__", with: version),
                                      injectionTime: .atDocumentStart, forMainFrameOnly: true)
            vue.configuration.userContentController.addUserScript(script)
        }
    }

    func detacher() {
        vue?.configuration.userContentController.removeScriptMessageHandler(forName: Pont.nom)
        minuterieNiveau?.invalidate()
    }

    var captureEnCours: Bool { enregistreur.enCours }

    // MARK: - Réception

    func userContentController(_ controller: WKUserContentController, didReceive message: WKScriptMessage) {
        // Seule notre page parle au pont : une caisse de paiement embarquée
        // n'obtient rien.
        let hote = message.frameInfo.securityOrigin.host.lowercased()
        guard hote == "lerapporteur.com" || hote == "www.lerapporteur.com" else { return }
        guard let corps = message.body as? [String: Any],
              let id = corps["id"] as? Int,
              let action = corps["action"] as? String else { return }
        let params = corps["params"] as? [String: Any] ?? [:]

        switch action {
        case "demarrer":
            enregistreur.demarrer { resultat in
                DispatchQueue.main.async {
                    switch resultat {
                    case .success:
                        self.lancerLeNiveau()
                        self.repondre(id, ["ok": true])
                    case .failure(let erreur):
                        self.repondre(id, ["erreur": self.decrire(erreur)])
                    }
                }
            }
        case "pause":
            enregistreur.pause(); repondre(id, ["ok": true])
        case "reprendre":
            enregistreur.reprendre(); repondre(id, ["ok": true])
        case "extrait":
            fileEnvois.async {
                let donnees = self.enregistreur.extrait()
                DispatchQueue.main.async {
                    if let donnees = donnees {
                        self.envoyerParMorceaux(id, donnees) { self.repondre(id, ["ok": true]) }
                    } else {
                        self.repondre(id, ["ok": true]) // rien à juger : la page n'enverra pas de sonde
                    }
                }
            }
        case "arreter":
            minuterieNiveau?.invalidate()
            fileEnvois.async {
                let url = self.enregistreur.arreter()
                let donnees = url.flatMap { try? Data(contentsOf: $0) }
                DispatchQueue.main.async {
                    guard let donnees = donnees, !donnees.isEmpty else {
                        self.repondre(id, ["erreur": "fichier introuvable"]); return
                    }
                    self.envoyerParMorceaux(id, donnees) {
                        self.repondre(id, ["ok": true, "octetsTotal": donnees.count])
                        // La page tient désormais le fichier (file d'attente) :
                        // l'exemplaire temporaire n'a plus de raison d'être.
                        if let url = url { try? FileManager.default.removeItem(at: url) }
                    }
                }
            }
        case "notifier":
            Notifications.partagees.poser(titre: params["titre"] as? String ?? "Rapporteur",
                                          texte: params["texte"] as? String ?? "")
            repondre(id, ["ok": true])
        case "enregistrementDemarre", "enregistrementTermine":
            // Le pendant du service Android : ici, c'est la session audio
            // active qui tient l'application éveillée — rien à faire de plus.
            repondre(id, ["ok": true])
        default:
            repondre(id, ["erreur": "action inconnue"])
        }
    }

    // MARK: - Réponses

    private func repondre(_ id: Int, _ reponse: [String: Any]) {
        guard let json = try? JSONSerialization.data(withJSONObject: reponse),
              let texte = String(data: json, encoding: .utf8) else { return }
        vue?.evaluateJavaScript("window.rapporteurIOS && window.rapporteurIOS._repondre(\(id), \(texte));")
    }

    /// Le binaire passe en base64, par tranches d'un mégaoctet, l'une après
    /// l'autre : une seule chaîne de 60 Mo étranglerait la vue web.
    private func envoyerParMorceaux(_ id: Int, _ donnees: Data, _ fini: @escaping () -> Void) {
        let pas = 1_048_576
        var debut = 0
        func suivant() {
            guard debut < donnees.count else { fini(); return }
            let fin = min(debut + pas, donnees.count)
            let b64 = donnees.subdata(in: debut..<fin).base64EncodedString()
            debut = fin
            vue?.evaluateJavaScript("window.rapporteurIOS && window.rapporteurIOS._morceau(\(id), \"\(b64)\");") { _, _ in
                suivant()
            }
        }
        suivant()
    }

    /// Le vu-mètre : dix relevés par seconde, poussés dans la page.
    private func lancerLeNiveau() {
        minuterieNiveau?.invalidate()
        minuterieNiveau = Timer.scheduledTimer(withTimeInterval: 0.1, repeats: true) { [weak self] _ in
            guard let self = self, self.enregistreur.enCours else { return }
            let n = self.enregistreur.niveau
            self.vue?.evaluateJavaScript("window.rapporteurIOS && (window.rapporteurIOS.capture.niveau = \(n));")
        }
    }

    private func decrire(_ erreur: Error) -> String {
        if let e = erreur as? Enregistreur.Erreur {
            switch e {
            case .microRefuse: return "micro refusé"
            case .dejaEnCours: return "déjà en cours"
            case .rienAEnregistrer: return "rien à enregistrer"
            case .moteur(let detail): return "moteur : \(detail)"
            }
        }
        return erreur.localizedDescription
    }
}

File: Rapporteur/Ressources/pont.js
/* Le pont, côté page — injecté par l'application au tout début du chargement,
   dans le cadre principal seulement.

   window.rapporteurIOS reprend l'interface du pont Android (version,
   plateforme, boutique, enregistrementDemarre/Termine) et y ajoute `capture`,
   la capture native : WebKit coupe le micro d'une page dès que l'écran
   s'éteint, l'application enregistre donc elle-même et rend le fichier ici.
   window.rapporteurAndroid en est l'ALIAS : pour le site, il veut dire
   « application mobile » — ses branches Android valent, et il lit
   rapporteurIOS là où l'iPhone diffère.

   Pas de ouvrirReunion : sur iPhone, une réunion tenue dans une autre
   application garde le micro pour elle (le système l'impose) — la salle ne
   propose donc pas d'ouvrir la réunion ici. */
(function () {
  "use strict";
  if (window.rapporteurIOS) { return; }

  var attentes = {};
  var suivant = 1;

  function appeler(action, params) {
    return new Promise(function (resolve, reject) {
      var id = suivant++;
      attentes[id] = { resolve: resolve, reject: reject, morceaux: [] };
      try {
        window.webkit.messageHandlers.rapporteur.postMessage({ id: id, action: action, params: params || {} });
      } catch (e) {
        delete attentes[id];
        reject(e);
      }
    });
  }

  function versOctets(b64) {
    var bin = atob(b64);
    var out = new Uint8Array(bin.length);
    for (var i = 0; i < bin.length; i++) { out[i] = bin.charCodeAt(i); }
    return out;
  }

  var pont = {
    version: function () { return "__VERSION__"; },
    plateforme: function () { return "ios"; },
    /* « appstore » : comme « play », les achats se font hors de l'application
       (règle 3.1.1 de l'App Store) — le site masque prix et caisses. */
    boutique: function () { return "appstore"; },
    enregistrementDemarre: function () { appeler("enregistrementDemarre"); },
    enregistrementTermine: function () { appeler("enregistrementTermine"); },
    notifier: function (titre, texte) { appeler("notifier", { titre: titre, texte: texte }); },

    capture: {
      /* Poussé par l'application dix fois par seconde, à l'échelle du vu-mètre. */
      niveau: 0,
      demarrer: function () { return appeler("demarrer"); },
      pause: function () { return appeler("pause"); },
      reprendre: function () { return appeler("reprendre"); },
      /* Les trois dernières minutes en .m4a, ou null s'il n'y a rien à juger. */
      extrait: function () {
        return appeler("extrait").then(function (r) {
          return r && r.octets && r.octets.length ? new Blob(r.octets, { type: "audio/mp4" }) : null;
        });
      },
      /* Le fichier de la séance, en UNE partie : la liste que la page attend. */
      arreter: function () {
        return appeler("arreter").then(function (r) {
          return r && r.octets && r.octets.length ? [new Blob(r.octets, { type: "audio/mp4" })] : [];
        });
      }
    },

    _morceau: function (id, b64) {
      var a = attentes[id];
      if (a) { a.morceaux.push(versOctets(b64)); }
    },
    _repondre: function (id, reponse) {
      var a = attentes[id];
      if (!a) { return; }
      delete attentes[id];
      if (reponse && reponse.erreur) { a.reject(new Error(reponse.erreur)); return; }
      reponse = reponse || {};
      if (a.morceaux.length) { reponse.octets = a.morceaux; }
      a.resolve(reponse);
    }
  };

  Object.defineProperty(window, "rapporteurIOS", { value: pont, writable: false, configurable: false });
  Object.defineProperty(window, "rapporteurAndroid", { value: pont, writable: false, configurable: false });
})();
